Flaky test: ServerMainShowPartitionsTest.testServerMainShowPartitions()

### Describe the bug

Error in log:
```
2023-10-04T21:05:01.5288150Z testServerMainShowPartitions[WITH_WAL-null](io.questdb.test.ServerMainShowPartitionsTest)  Time elapsed: 10.554 sec  <<< FAILURE!
2023-10-04T21:05:01.5288950Z java.lang.AssertionError: expected:<count
2023-10-04T21:05:01.5289390Z 1000000
2023-10-04T21:05:01.5289710Z > but was:<count
2023-10-04T21:05:01.5290190Z 0
2023-10-04T21:05:01.5290510Z >
2023-10-04T21:05:01.5290850Z 	at org.junit.Assert.fail(Assert.java:89)
2023-10-04T21:05:01.5291670Z 	at org.junit.Assert.failNotEquals(Assert.java:835)
2023-10-04T21:05:01.5292080Z 	at org.junit.Assert.assertEquals(Assert.java:120)
2023-10-04T21:05:01.5292480Z 	at io.questdb.test.tools.TestUtils.assertEquals(TestUtils.java:362)
2023-10-04T21:05:01.5292900Z 	at io.questdb.test.tools.TestUtils.assertEquals(TestUtils.java:345)
2023-10-04T21:05:01.5293330Z 	at io.questdb.test.tools.TestUtils.assertSql(TestUtils.java:643)
2023-10-04T21:05:01.5293760Z 	at io.questdb.test.ServerMainShowPartitionsTest.waitForData(ServerMainShowPartitionsTest.java:200)
2023-10-04T21:05:01.5294280Z 	at io.questdb.test.ServerMainShowPartitionsTest.lambda$testServerMainShowPartitions$2(ServerMainShowPartitionsTest.java:128)
```

Full log:
[206.txt](https://github.com/questdb/questdb/files/12810067/206.txt)


### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**: Mac (mac-other CI)
- **Browser**:
```


### Additional context

_No response_

--- a/core/src/test/java/io/questdb/test/ServerMainShowPartitionsTest.java
+++ b/core/src/test/java/io/questdb/test/ServerMainShowPartitionsTest.java
@@ -37,20 +37,18 @@ import io.questdb.log.LogFactory;
 import io.questdb.mp.SOCountDownLatch;
 import io.questdb.std.LongList;
 import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
 import io.questdb.std.Os;
 import io.questdb.std.str.StringSink;
 import io.questdb.test.cairo.TableModel;
 import io.questdb.test.tools.TestUtils;
-import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.List;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -130,9 +128,9 @@ public class ServerMainShowPartitionsTest extends AbstractBootstrapTest {
 
                 int numThreads = 5;
                 SOCountDownLatch completed = new SOCountDownLatch(numThreads);
-                AtomicReference<List<Throwable>> errors = new AtomicReference<>(new ArrayList<>());
-                List<SqlCompiler> compilers = new ArrayList<>(numThreads);
-                List<SqlExecutionContext> contexts = new ArrayList<>(numThreads);
+                AtomicReference<Throwable> errors = new AtomicReference<>();
+                ObjList<SqlCompiler> compilers = new ObjList<>(numThreads);
+                ObjList<SqlExecutionContext> contexts = new ObjList<>(numThreads);
                 for (int i = 0; i < numThreads; i++) {
                     SqlCompiler compiler = qdb.getEngine().getSqlCompiler();
                     SqlExecutionContext context = createSqlExecutionCtx(qdb.getEngine());
@@ -142,26 +140,24 @@ public class ServerMainShowPartitionsTest extends AbstractBootstrapTest {
                         try {
                             assertShowPartitions(finallyExpected, tableToken, compiler, context);
                         } catch (Throwable err) {
-                            errors.get().add(err);
+                            errors.compareAndSet(null, err);
                         } finally {
                             completed.countDown();
                         }
                     }).start();
                 }
-                if (!completed.await(TimeUnit.SECONDS.toNanos(3L))) {
+                if (!completed.await(TimeUnit.MINUTES.toNanos(1))) {
+                    errors.compareAndSet(null, new AssertionError("Timed out waiting for threads to complete"));
                     TestListener.dumpThreadStacks();
                 }
                 dropTable(defaultCompiler, defaultContext, tableToken);
-                for (int i = 0; i < numThreads; i++) {
-                    compilers.get(i).close();
-                    contexts.get(i).close();
-                }
-                compilers.clear();
-                contexts.clear();
+                Misc.freeObjListAndClear(compilers);
+                Misc.freeObjListAndClear(contexts);
 
                 // fail on first error found
-                for (Throwable t : errors.get()) {
-                    Assert.fail(t.getMessage());
+                Throwable firstError = errors.get();
+                if (firstError != null) {
+                    throw new AssertionError(firstError);
                 }
             }
         });

--- a/core/src/test/java/io/questdb/test/ServerMainShowPartitionsTest.java
+++ b/core/src/test/java/io/questdb/test/ServerMainShowPartitionsTest.java
@@ -38,7 +38,6 @@ import io.questdb.mp.SOCountDownLatch;
 import io.questdb.std.LongList;
 import io.questdb.std.Misc;
 import io.questdb.std.ObjList;
-import io.questdb.std.Os;
 import io.questdb.std.str.StringSink;
 import io.questdb.test.cairo.TableModel;
 import io.questdb.test.tools.TestUtils;
@@ -120,8 +119,7 @@ public class ServerMainShowPartitionsTest extends AbstractBootstrapTest {
                 CairoConfiguration cairoConfig = qdb.getConfiguration().getCairoConfiguration();
 
                 TableToken tableToken = createPopulateTable(cairoConfig, engine, defaultCompiler, defaultContext, tableName);
-                // wait for the rows to end up in the table
-                waitForData(tableName, defaultCompiler, defaultContext);
+                engine.awaitTable(tableName, 30, TimeUnit.SECONDS);
 
                 String finallyExpected = replaceSizeToMatchOS(EXPECTED, dbPath, tableToken.getTableName(), engine);
                 assertShowPartitions(finallyExpected, tableToken, defaultCompiler, defaultContext);
@@ -182,23 +180,6 @@ public class ServerMainShowPartitionsTest extends AbstractBootstrapTest {
                 cursor0.toTop();
                 AbstractCairoTest.assertCursor(finallyExpected, false, true, false, cursor1, meta, sink, rows, false);
                 cursor1.toTop();
-            }
-        }
-    }
-
-    private static void waitForData(String tableName, SqlCompiler defaultCompiler, SqlExecutionContext defaultContext) throws SqlException {
-        long time = System.currentTimeMillis();
-        StringSink sink = new StringSink();
-        while (true) {
-            try {
-                TestUtils.assertSql(defaultCompiler, defaultContext, "select count() from " + tableName, sink, "count\n" +
-                        "1000000\n");
-                break;
-            } catch (AssertionError e) {
-                if (System.currentTimeMillis() - time > 5000) {
-                    throw e;
-                }
-                Os.sleep(5);
             }
         }
     }

